transport/disorder/stream_dialer.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package disorder
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"syscall"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+)
+
+var defaultTTL = 64
+
+type disorderDialer struct {
+	dialer     transport.StreamDialer
+	splitPoint int64
+}
+
+var _ transport.StreamDialer = (*disorderDialer)(nil)
+
+// NewStreamDialer creates a [transport.StreamDialer] that splits the outgoing stream after writing "prefixBytes" bytes
+// using [SplitWriter].
+func NewStreamDialer(dialer transport.StreamDialer, prefixBytes int64) (transport.StreamDialer, error) {
+	if dialer == nil {
+		return nil, errors.New("argument dialer must not be nil")
+	}
+	return &disorderDialer{dialer: dialer, splitPoint: prefixBytes}, nil
+}
+
+// DialStream implements [transport.StreamDialer].DialStream.
+func (d *disorderDialer) DialStream(ctx context.Context, remoteAddr string) (transport.StreamConn, error) {
+	innerConn, err := d.dialer.DialStream(ctx, remoteAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// this strategy only works when we set TCP as a strategy
+	tcpConn, ok := innerConn.(*net.TCPConn)
+	if !ok {
+		return nil, fmt.Errorf("split strategy only works with direct TCP connections")
+	}
+
+	err = tcpConn.SetNoDelay(true)
```

Comment:
Isn't this the default? Also, I'd rather not silently change settings. The base dialer provider should be responsible for that.

Perhaps we check if it's set and error if not. Though I don't see an easy way to check.

Add to the comments that the TCPConn must have NoDelay set.

Reply:
> Isn't this the default?

Honestly no idea. In Linux it's default now (but only for recent kernels, most android likely will need years to adopt the change), but in golang no idea.

I've added this in an attempt to debug something, but the issue was somewhere else, so this is no longer needed

Reply:
It's enabled by default in Go, so no need to set it: https://pkg.go.dev/net#TCPConn.SetNoDelay

Reply:
awesome

Reply:
fixed